utils/wasm-builder/src/builder.rs

```diff
@@ -0,0 +1,245 @@
+/// 2. Select the project to build using the methods of [`WasmBuilderSelectProject`].
+/// 3. Set additional `RUST_FLAGS` or a different name for the file containing the WASM code
+///    using methods of [`WasmBuilder`].
+/// 4. Build the WASM binary using [`Self::build`].
+pub struct WasmBuilder {
+	/// Flags that should be appended to `RUST_FLAGS` env variable.
+	rust_flags: Vec<String>,
+	/// The name of the file that is being generated in `OUT_DIR`.
+	///
+	/// Defaults to `wasm_binary.rs`.
+	file_name: Option<String>,
+	/// The path to the `Cargo.toml` of the project that should be build
+	/// for wasm.
+	project_cargo_toml: PathBuf,
+}
+
+impl WasmBuilder {
+	/// Create a new instance of the builder.
+	pub fn new() -> WasmBuilderSelectProject {
+		WasmBuilderSelectProject {
+			_ignore: (),
+		}
+	}
+
+	/// Enable exporting `__heap_base` as global variable in the WASM binary.
+	///
+	/// This adds `-Clink-arg=--export=__heap_base` to `RUST_FLAGS`.
+	pub fn export_heap_base(mut self) -> Self {
+		self.rust_flags.push("-Clink-arg=--export=__heap_base".into());
+		self
+	}
+
+	/// Set the name of the file that will be generated in `OUT_DIR`.
+	///
+	/// This file needs to be included to get access to the build WASM binary.
+	///
+	/// If this function is not called, `file_name` defaults to `wasm_binary.rs`
+	pub fn set_file_name(mut self, file_name: impl Into<String>) -> Self {
+		self.file_name = Some(file_name.into());
+		self
+	}
+
+	/// Instruct the linker to import the memory into the WASM binary.
+	///
+	/// This adds `-C link-arg=--import-memory` to `RUST_FLAGS`.
+	pub fn import_memory(mut self) -> Self {
+		self.rust_flags.push("-C link-arg=--import-memory".into());
+		self
+	}
+
+	/// Append the given `flag` to `RUST_FLAGS`.
+	///
+	/// `flag` is appended as is, so it needs to be a valid flag.
+	pub fn append_to_rust_flags(mut self, flag: impl Into<String>) -> Self {
+		self.rust_flags.push(flag.into());
+		self
+	}
+
+	/// Build the WASM binary.
+	pub fn build(self) {
+		let out_dir = PathBuf::from(env::var("OUT_DIR").expect("`OUT_DIR` is set by cargo!"));
+		let file_path = out_dir.join(self.file_name.unwrap_or_else(|| "wasm_binary.rs".into()));
+
+		if check_skip_build() {
+			// If we skip the build, we still want to make sure to be called when an env variable
+			// changes
+			generate_rerun_if_changed_instructions();
+
+			provide_dummy_wasm_binary_if_not_exist(&file_path);
+
+			return;
+		}
+
+		build_project(
+			file_path,
+			self.project_cargo_toml,
+			self.rust_flags.into_iter().map(|f| format!("{} ", f)).collect(),
+		);
+
+		// As last step we need to generate our `rerun-if-changed` stuff. If a build fails, we don't
+		// want to spam the output!
+		generate_rerun_if_changed_instructions();
+	}
+}
+
+/// Generate the name of the skip build environment variable for the current crate.
+fn generate_crate_skip_build_env_name() -> String {
+	format!(
+		"SKIP_{}_WASM_BUILD",
```

Comment:
This is an hidden feature, only for people who read the source 🙈

Reply:
I propose that you read this: https://github.com/paritytech/substrate/blob/79a1d10988ff068d1b686f8370f97469f3e2c7de/utils/wasm-builder/README.md#environment-variables

:P

Reply:
I just looked at the list and didn't check the last part 🙈